### ISSUE TITLE
Terraform deploy plugin

### ISSUE THREAD

**Issue by issacg** [hashicorp/waypoint#630]
I'm surprised it hasn't bit brought up yet, but a generic plugin to call Terraform would be a great addition.

Some of the plugins, like aws-ec2, are very opinionated about the underlying infrastructure, but don't really allow tunables beyond what the plugin author feels is important.

Many admins have already finely tuned setups written in Terraform, and I would imagine it would be trivial-ish to allow a plugin to point to a directory on the filesystem and maybe a workspace (or, down the line, a TF Cloud hosted workspace), and a map of variables to plug in to terraform instead.  This would allow the ultimate in tuning for the underlying infrastructure.

Something like

```hcl
deploy {
  use "terraform" {
    path = "/path/to/terraform",
    workspace = "default",
    variables = {
       foo = "bar"
       port = 80
    }
    // Maybe environment variable too?
 }
}
```

The plugin would terraform init, terraform get -update and finally terraform apply.  The state could use an existing storage.  (Perhaps down the line waypoint server would include a remote terraform storage to be stored per app/workspace?)

Kudos on the release @mitchellh ! Looking good, so far :)

**Comment by mitchellh** [comment]
Yep! This is something we want to do but aren't quite sure how we want to do it yet. We know conceptually it makes sense but we need to think about this a bit more. Thank you for the example though that does help

**Comment by issacg** [comment]
Well, to share my user-oriented perspective (just as a sounding board for you), it's not clear to me how you (Hashicorp) sees the "deploy" abstraction working.  

At the moment you have the EC2 plugin, which is good for AMI-based deploys, the netlify plugin but I'd never heard of netlify until I saw it in the docs, and everything else is docker.  

I don't really have a strong opinion about terraform beyond it intuitively feeling it belongs there, and it being a huge improvement over the EC2 plugin you have today.  Tuning the EC2 plugin needs to be there (the first thing I thought of was "How do I ensure my root filesystem is on encrypted EBS?" which wasn't in the current plugin) but allowing everything to be tunable would just turn it into a mirror of the existing Terraform drivers which seems like 1) duplicated effort and 2) over-complexity for the Waypoint system which, to my understanding, is supposed to "just work" with as minimal amount of config as can reasonably be possible.

**Comment by rugwirobaker** [comment]
> Waypoint system which, to my understanding, is supposed to "just work" with as minimal amount of config as can reasonably be possible.

This really hits home. It would ruin the experience if the configuration ended up going kilometres of lines which it might as the number of plugin options increase.

**Comment by curtbushko** [comment]
Terraform is difficult because it doesn't exactly fit in with how the waypoint server runs. All of the current plugins are calling  some external API. ECS, Google Cloud Run, etc are all running somewhere else. Even docker is "external" to the waypoint server. Terraform execution is 'local'.

I am thinking that the way to make this work is to use docker in docker to run an execution of a worker container that runs terraform. You already have the --platform option for waypoint server so that it knows where it is running. It can use that to determine what kind of worker to run (docker image vs kubernetes pod/job). You'd pipe the stdout from the worker into waypoint cli.

One option that would be required is 'image' so that you could let the user do whatever they want with terraform. They could create their own image, cache the providers they use, credentials, etc.

```terraform
deploy {
  use "terraform" {
    image = "hashicorp/terraform:0.14.0",
    path = "/path/to/terraform",
    workspace = "default",
    variables = {
      foo = "bar"
      port = 80
     }
  }
}
```

Note: I haven't dug into the code yet to see if you are doing this yet but I assume you are just using channels instead of dind.

**Comment by issacg** [comment]
You'd need something like that `image` anyway, because you'll have users of Terraform Cloud and users of Terraform Enterprise, in which case you'd need to supply a server and workspace at the least.  

It wasn't my intent that my config example be complete - for example, another thing which I intentionally omitted is some variable pointing to the "artifact" - a link to a ZIP file (for serverless), an AMI id (for EC2 and friends), a docker server/repository:tag (for docker), etc

**Comment by curtbushko** [comment]
It almost feels like you could get into terragrunt functionality with how it bootstraps terraform.

**Comment by Vlaaaaaaad** [comment]
> Terraform is difficult because it doesn't exactly fit in with how the waypoint server runs. All of the current plugins are calling some external API. ECS, Google Cloud Run, etc are all running somewhere else. Even docker is "external" to the waypoint server. Terraform execution is 'local'.

Could that function be fulfilled by something like say [Atlantis](https://www.runatlantis.io)? Or Terraform Cloud?

**Comment by tehmaspc** [comment]
This would be an awesome feature to have.

Just to give you our perspective...

We currently build "Pipelines" for our business. Each Pipeline facilitates the delivery of applications on diff private/public cloud platforms. Where necessary (e.g. AWS) - part of the Pipeline's responsibility is to deploy the necessary infrastructure required by the application(s) via Terraform for that environment (i.e dev, test, stag, prod). Thus, we deploy our applications and infrastructure per environment and call out to Terraform in our Pipeline to do this very thing. We basically do what @curtbushko detailed above - call a versioned TF container and stuff the state into the proper workspace for that app per environment.

Much of the abstractions I see in some of the Waypoint plugins - e.g. Lambda and ECS - we have already built but w/ our own code. I would love to leverage Waypoint and reduce our internal code burden but without a Terraform infrastructure deployment model we would have to supplement to this model, thus having the ability to deploy an applications infrastructure needs (via TF) alongside the Waypoint execution would be an interesting model for us.

Cheers!

**Comment by acherifi** [comment]
That would be an awesome feature to have since our workflow needs to create some Cloud buckets at deploym we could run Terraform to create the bucket just before deploying our application.
That would allow us to deploy our code and its dependencies without changing anything to our process

**Comment by denzuko** [comment]
To add a little into the discussion on the "how we want to do" part. The current topic of building an OCI image for terraform and potentially the tfplan, hcl, etc. is an interesting idea; something we already do over here at @Daplanet since it's portable enough to run in either github actions or cloud functions. But we've been looking into Waypoint, nomad, packer, and sentinel to cut down on and/or simplify the amount of technical maintenance overhead.

So seeing how most hashicorp tools can seamlessly interoperate with each other (e.g. nomad can use consul and vault but does not need to, waypoint can use both nomad, consul, vault, while each are able to run sentinel given its in the path). It would be nice to have that kind of non dependant interoperability also with boundary, terraform, and packer.

An example would be just like with sentinel, waypoint  could check for the PATH for terraform then offer that as a provisioner. From the HCL it could just be another value for the use stanza.

**Comment by yuriy-yarosh** [comment]
I do strongly agree with @denzuko
Waypoint server should be able sync it's internal state with terraform, and reuse the existing terraform locks.
Waypoint operations have to be atomic, versionable and revertable the same way as terraform ones.

Ideally, Waypoint project configs should be embedable into terraform modules directly, with a provider, which would check version drift and pass the respective project/workspace config to the server. Warn if versions do differ, and ask for a proper amend.

I don't really like a single `waypoint.hcl` file and that there's no clear way to specify multi-project dependencies, like terraform does with `depends_on` meta-argument.

But once again, if we could embed waypoint into terraform with a provider directly - we won't need to reinvent that wheel.

The other thing is that building a docker container that is reused across numerous projects is commonplace, alongside templating Dockerfile's... and I'm not sure how build artifacts should be shared for a multiproject setup, but again, it's something easily doable with a custom terraform provider and a couple additions to the existing waypoint config spec.

I think that tracking waypoint dependencies with terraform and locking the state there is a way to go.

**Comment by jcampbell05** [comment]
For our use case we aren't sure we would want full fat terraform having used it before it would be over kill for our project. Instead we would love a plugin for various cloud providers which provided various primitives similar to our current use of serverless-lyft (https://www.npmjs.com/package/serverless-lift)

Waypoint to us is a much closer replacement for Serverless. Both want to expose as little of the infrastructure to you as possible and automatically generate what is needed for your application to run.

It seems at the moment Waypoint's eco-system of plugins is too limited for us to migrate over. Ideally eventually would love to just specify we want a queue and the AWS Waypoint plugin would setup the SQS queue and a trigger which would trigger our Application.
